### PR TITLE
Enh ssl module

### PR DESCRIPTION
**Pull request by twerthi**
Replaces #263 by placing the functionality into new resource.  @matt-richardson had asked the question in #263 regarding whether there was a PowerShell equivalent for the `netsh http` usage.  The only thing I was able to find was https://docs.microsoft.com/en-us/powershell/module/networktransition/add-netiphttpscertbinding?view=win10-ps.  This cmdlet seems to be only for IP-HTTPS which I didn't believe was the same thing, please correct me if I'm wrong.

From documentation,

```
This cmdlet is similar to the netsh http add sslcert command, but the cmdlet should only be used for IP-HTTPS.
```

### CODE REVIEW

File: OctopusDSC/DSCResources/cOctopusServerSslCertificate/cOctopusServerSslCertificate.psm1 [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+        [ValidateSet("Present", "Absent")]
+        [string] $Ensure = "Present",
+        [ValidateNotNullOrEmpty()]
+        [Parameter(Mandatory)]
+        [string] $Thumbprint,
+        [ValidateNotNullOrEmpty()]
+        [Parameter(Mandatory)]
+        [string] $StoreName,
+        [ValidateNotNullOrEmpty()]
+        [string] $Port = "443"
```
Comment (matt-richardson) [inline OctopusDSC/DSCResources/cOctopusServerSslCertificate/cOctopusServerSslCertificate.psm1:124]: Make this an int

File: OctopusDSC/DSCResources/cOctopusServerSslCertificate/cOctopusServerSslCertificate.psm1 [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+        else {
+            Write-Verbose "Configuration parameter '$key' matches the requested value '$requestedValue'"
+        }
+    }
+
+    return $currentConfigurationMatchesRequestedConfiguration
+}
+
+function Set-TargetResource
+{
```
Comment (matt-richardson) [inline OctopusDSC/DSCResources/cOctopusServerSslCertificate/cOctopusServerSslCertificate.psm1:111]: Can you pull opening braces onto the line above to throughout the file for consistency?

File: OctopusDSC/Tests/cOctopusServerActiveDirectoryAuthentication.Tests.ps1 [hunk 3]
```diff
@@ -105,6 +105,7 @@ try
 
             Context 'Set-TargetResource' {
                 BeforeAll {
+                    $dscHelpersPath = Resolve-Path "$PSCommandPath/../../OctopusDSCHelpers.ps1"
```
Comment (matt-richardson) [inline OctopusDSC/Tests/cOctopusServerActiveDirectoryAuthentication.Tests.ps1:108]: This doesn't feel right. If anything, if it needs changing, the line below should reference `$script:dscHelpersPath` instead.
If it does need to change, it should be in a separate PR, because its unrelated to this change.

File: OctopusDSC/Tests/cOctopusServerSslCertificate.Tests.ps1 [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+                        }
+                    }
+
+                    $result = Get-TargetResource -InstanceName "OctopusServer" `
+                                                 -Ensure "Present" `
+                                                 -Thumbprint "dfcbdb879e5315e05982c05793e57c39241797e3" `
+                                                 -StoreName "My"
+
+                    $result.Ensure | Should -Be 'Present'
+                    $result.Port | Should -Be "443"
```
Comment (matt-richardson) [inline OctopusDSC/Tests/cOctopusServerSslCertificate.Tests.ps1:37]: Check the other values as well

File: OctopusDSC/Tests/cOctopusServerSslCertificate.Tests.ps1 [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+{
+    $prefix = [guid]::NewGuid().Guid -replace '-'
+    $module = Import-Module $modulePath -Prefix $prefix -PassThru -ErrorAction Stop
+
+    InModuleScope $module.Name {
+        $sampleConfigPath = Split-Path $PSCommandPath -Parent
+        $sampleConfigPath = Join-Path $sampleConfigPath "SampleConfigs"
+
+        Describe 'cOctopusServerSslCertificate' {
+            Context 'Get-TargetResource - when present and port not specified' {
```
Comment (matt-richardson) [inline OctopusDSC/Tests/cOctopusServerSslCertificate.Tests.ps1:20]: Should be another `Describe` block for `Get-TargetResource`, and you can then remove that from the name of the context blocks.

**Review by matt-richardson — CHANGES_REQUESTED** [review]
Nice work - coming along really nicely.
I've added a few comments, mainly about tidying up, but a few larger ones.
I'd also like to see some e2e tests (in the scenario files) to prove it works in the "real world".
Can you also create a readme for this, and reference it from the main readme?

File: OctopusDSC/Tests/cOctopusServerActiveDirectoryAuthentication.Tests.ps1 (hunk 3 above)
Comment (twerthi) [reply OctopusDSC/Tests/cOctopusServerActiveDirectoryAuthentication.Tests.ps1]: Removed to PR #268

File: OctopusDSC/Tests/cOctopusServerSslCertificate.Tests.ps1 (hunk 5 above)
Comment (twerthi) [reply OctopusDSC/Tests/cOctopusServerSslCertificate.Tests.ps1:21]: I was following the pattern in cOctopusServerSpace.Tests.ps1, is that one also incorrect?

Comment (matt-richardson) [reply OctopusDSC/Tests/cOctopusServerSslCertificate.Tests.ps1:22]: It's not so much "incorrect", as "I'd say my suggested approach here is slightly better"

File: README-cOctopusServerSslCertificate.md [hunk 6]
```diff
@@ -0,0 +1,39 @@
+# README-cOctopusServerSslCertificate
+
```
Comment (matt-richardson) [inline README-cOctopusServerSslCertificate.md:2]: Be good to add a description of what it does and when you'd use it.

File: README-cOctopusServerSslCertificate.md [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+Configuration SampleConfig
+{
+    Import-DscResource -Module OctopusDSC
+
+    Node "localhost"
+    {
+        cOctopusServerSslCertificate SSLCert
+        {
+            InstanceName = "OctopusServer"
+            Thumbprint = "c42a148bcd3959101f2e7a3d76edb924bff84b6b"
```
Comment (matt-richardson) [inline README-cOctopusServerSslCertificate.md:17]: Might be worth a comment to say where to get this from

File: README-cOctopusServerSslCertificate.md [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+Start-DscConfiguration .\SampleConfig -Verbose -wait
+
+Test-DscConfiguration
+```
+
+## Properties
+| Property                                  | Type         | Default Value    | Description |
+| ------------------------------------------| ------------ | -----------------| ------------|
+| `Ensure`                              | `string` - `Present` or `Absent`                    | `Present`                                                       | The desired state of the Octopus Server - effectively whether to install or uninstall. |
```
Comment (matt-richardson) [inline README-cOctopusServerSslCertificate.md:35]: My OCD humbly requests that you line these up :)

File: OctopusDSC/Tests/cOctopusServerSslCertificate.Tests.ps1 [hunk 9]
```diff
@@ -19,87 +17,84 @@ try
...
         Describe 'cOctopusServerSslCertificate' {
-            Context 'Get-TargetResource - when present and port not specified' {
+            Describe 'Get-TargetResource - when present and port not specified' {
                 It 'Returns present and 443 when SSL binding exists' {
                     Mock Get-CurrentSSLBinding {
                         return [PSCustomObject]@{
                             IPPort = "0.0.0.0:443"
-                            CertificateHash = $certificateThumbprint
-                            AppID = $octopusServerApplicationId
+                            CertificateHash = "dfcbdb879e5315e05982c05793e57c39241797e3"
```
Comment (matt-richardson) [inline OctopusDSC/Tests/cOctopusServerSslCertificate.Tests.ps1:25]: This looks like it undoes the previous change to move these into variables?

File: OctopusDSC/DSCResources/cOctopusServerSslCertificate/cOctopusServerSslCertificate.psm1 [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+        [string] $StoreName,
+        [ValidateNotNullOrEmpty()]
+        [int] $Port = 443
+    )
+
+    $existingSSLConfig = (Get-CurrentSSLBinding -ApplicationId $octopusServerApplicationId -Port $Port -IPAddress $IPAddress)
+
+    if ($null -ne $existingSSLConfig) {
+        $existingSSLPort = [int](($existingSSLConfig.IPPort).Split(":")[1])
+        $existingSSLThumbprint = $existingSSLConfig.CertificateHash.Trim()
```
Comment (matt-richardson) [inline OctopusDSC/DSCResources/cOctopusServerSslCertificate/cOctopusServerSslCertificate.psm1:50]: Bit weird we are using "thumbprint" and "hash" hee - would be good to make these consistent

**Review by matt-richardson — COMMENTED** [review]
A few comments

File: OctopusDSC/Tests/cOctopusServerSslCertificate.Tests.ps1 [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+
+                    $result.Ensure | Should -Be 'Present'
+                    $result.Port | Should -Be "443"
+                    $result.StoreName | Should -Be "My"
+                    $result.Thumbprint | Should -Be "dfcbdb879e5315e05982c05793e57c39241797e3"
+                }
+            }
+
+            Describe 'Get-TargetResource - when present and port specified' {
+                It 'Returns present and 1443 when SSL binding exists' {
```
Comment (matt-richardson) [inline OctopusDSC/Tests/cOctopusServerSslCertificate.Tests.ps1:44]:
```suggestion
                It 'Returns correct data when SSL binding on non standard port exists' {
```

File: OctopusDSC/Tests/cOctopusServerSslCertificate.Tests.ps1 [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+
+                    $result.Ensure | Should -Be 'Present'
+                    $result.Port | Should -Be "1443"
+                    $result.StoreName | Should -Be "My"
+                    $result.Thumbprint | Should -Be "dfcbdb879e5315e05982c05793e57c39241797e3"
+                }
+            }
+
+            Describe 'Get-TargetResource - when absent' {
+                It 'Returns absent when binding does not exist' {
```
Comment (matt-richardson) [inline OctopusDSC/Tests/cOctopusServerSslCertificate.Tests.ps1:68]:
```suggestion
                It 'Returns correct data when binding does not exist' {
```

File: OctopusDSC/Tests/cOctopusServerSslCertificate.Tests.ps1 [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+
+                    $result.Ensure | Should -Be "Absent"
+                    $result.StoreName | Should -Be $null
+                    $result.Thumbprint | Should -Be $null
+                    $result.Port | Should -Be $null
+                }
+            }
+
+            Describe 'Test-TargetResource - when present' {
+                It 'Returns false when port differs' {
```
Comment (matt-richardson) [inline OctopusDSC/Tests/cOctopusServerSslCertificate.Tests.ps1:86]: Do we need more tests here for:
* cert thumbprint is different
* certstore is different

File: README-cOctopusServerSslCertificate.md [hunk 14]
```diff
@@ -0,0 +1,41 @@
+# README-cOctopusServerSslCertificate
+
+This resource will bind a certificate identified by the StoreName and Thumbprint to the SSL entry in the `WebListenPrefix` element of the cOctopusServer resource.  The port defined for the SSL entry must match to bind properly.
```
Comment (matt-richardson) [inline README-cOctopusServerSslCertificate.md:3]: This felt too detailed - I've tried to re-write so that it's more obvious to an new-comer. What do you think?
```suggestion
This resource binds a SSL certificate to allow Octopus Server to listen over HTTPS.
It binds the certificate stored in the Windows Certificate store identified by the `StoreName` and `Thumbprint` to the port that the sever will listen on.
This port must match the https entry in the `WebListenPrefix` element of the cOctopusServer resource. 
```

File: README-cOctopusServerSslCertificate.md [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+```
+
+## Properties
+| Property                                  | Type         | Default Value    | Description |
+| ------------------------------------------| ------------ | -----------------| ------------|
+| `Ensure`                                  | `string` - `Present` or `Absent`| `Present`   | The desired state of the Octopus Server - effectively whether to install or uninstall. |
+| `InstanceName`                            | `string`     |                  | The name of the Octopus Server instance. Use `OctopusServer` by convention unless you have more than one instance. |
+| `Thumbprint`                              | `string`     |                  | Thumbprint of the SSL certificate to use with Octopus Deploy |
+| `StoreName`                               | `string` - `My` or `WebHosting` |          | Which certificate store to look for the thumbprint in. |
+| `Port`                                    | `int`        |                  | The port to use SSL on.  |
```
Comment (matt-richardson) [inline README-cOctopusServerSslCertificate.md:41]: My OCD is strong 😄:
```suggestion
| Property                                  | Type                            | Default Value    | Description                                                                                                        |
| ------------------------------------------| ------------------------------- | -----------------| ------------------------------------------------------------------------------------------------------------------ |
| `Ensure`                                  | `string` - `Present` or `Absent`| `Present`        | The desired state of the Octopus Server - effectively whether to install or uninstall.                             |
| `InstanceName`                            | `string`                        |                  | The name of the Octopus Server instance. Use `OctopusServer` by convention unless you have more than one instance. |
| `Thumbprint`                              | `string`                        |                  | Thumbprint of the SSL certificate to use with Octopus Deploy,                                                      |
| `StoreName`                               | `string` - `My` or `WebHosting` |                  | Which certificate store to look for the thumbprint in.                                                             |
| `Port`                                    | `int`                           |                  | The port to use SSL on.                                                                                            |
```

File: Tests/Scenarios/Server_Scenario_03_Remove.ps1 [hunk 16]
```diff
@@ -1,3 +1,28 @@
+function Get-CurrentSSLBinding {
```
Comment (matt-richardson) [inline Tests/Scenarios/Server_Scenario_03_Remove.ps1:1]: I'm not a huge fan of this being here, but it works... 
(Feels like it should be extracted somewhere else, like a shared module or something).
_Not saying we should change it as part of this PR though_

File: Tests/Scenarios/Server_Scenario_03_Remove.ps1 [hunk 17]
```diff
@@ -24,7 +58,7 @@ Configuration Server_Scenario_03_Remove
             Name = "OctopusServer"
 
             # The url that Octopus will listen on
-            WebListenPrefix = "http://localhost:81"
+            WebListenPrefix = "http://localhost:81,https://localhost"
```
Comment (matt-richardson) [inline Tests/Scenarios/Server_Scenario_03_Remove.ps1:61]: Do we even need this line? If we're removing the instance, then telling it what urls to listen on is a bit weird.

**Review by matt-richardson — APPROVED** [review]
Nice work, @twerthi! It's come together really nicely.
I'm happy with where this is at now, so I'll leave it with you to merge when you're happy.
I've added a few suggestions that can be directly committed & there's 2 comments (extra tests + a potential unnecessary line) that might be better to consider separately so we can get this shipped.
